app/src/main/java/ch/epfl/culturequest/SignUpActivity.java

```diff
@@ -20,35 +29,135 @@ public class SignUpActivity extends AppCompatActivity {
+            email = findViewById(R.id.editTextTextEmailAddress);
+            signUp = findViewById(R.id.sign_up_manually);
+            signIn = findViewById(R.id.sign_in_manually);
+            issues = findViewById(R.id.issues);
+            password.addTextChangedListener(passwordWatcher);
             // Sets the sign in launcher before activity goes to Resume state
             Authenticator.setSignInLauncher(this);
             findViewById(R.id.sign_in_button).setOnClickListener(k -> Authenticator.signIn(this));
         }
-
         // Otherwise directly signIn
         else {
             Authenticator.signIn(this);
         }
     }
 
+
+    @SuppressLint("SetTextI18n")
+    public void signUp(View v) {
+        String pw = password.getText().toString();
+        if (!hasDigit(pw)) {
+            issues.setText("Password must contain at least 1 digit");
+            return;
+        }
+        if (!hasSpecialChar(pw)) {
+            issues.setText("Password must contain at least 1 special character");
+            return;
+        }
+        if (pw.length() < 8) {
+            issues.setText("Password should be minimum 8 characters");
+            return;
+        }
+
+        if (pw.matches(PW_REGEX)) {
+            Authenticator.manualSignUp(email.getText().toString(), password.getText().toString()).thenCompose((success) -> {
+                if (!success.get()) {
+                    issues.setText("This email is already used");
+                    return null;
+                } else {
+                    return Authenticator.manualSignIn(email.getText().toString(), password.getText().toString());
+                }
+            }).thenCompose(success2 -> {
+                if (success2.get()) {
+                    return Authenticator.signIn(this);
+                }
+                return null;
+            });
+        }
+    }
+
+    @SuppressLint("SetTextI18n")
+    public void signIn(View v) {
+        if (password.getText().toString().matches(PW_REGEX)) {
+            Authenticator.manualSignIn(email.getText().toString(), password.getText().toString()).handle((exists, t) -> {
+                if (t != null) t.printStackTrace();
+                else if (!exists.get()) {
+                    issues.setText("Wrong sign in credentials");
+                } else {
+                    Authenticator.signIn(this);
+                }
+                return null;
+            });
+        } else {
+            password.setClickable(false);
+        }
+    }
+
     @Override
     protected void onResume() {
         super.onResume();
-        try{
+        try {
             TournamentManagerApi.handleTournaments(this);
         } catch (Exception e) {
             e.printStackTrace();
         }
     }
+
+    private void setEnabled(boolean enabled) {
+        float alpha = enabled ? 1f : 0.3f;
+        signUp.setAlpha(alpha);
+        signIn.setAlpha(alpha);
+        signIn.setClickable(enabled);
```

Comment:
Why only SignIn button is set clickable and not the SignUp one too ?

Reply:
this way if a user clicks on signup they get a message to tell them why their password or email is not valid

Reply:
oh ok